server/cluster_worker.go

```diff
@@ -294,352 +35,212 @@ func (c *raftCluster) handleAddPeerReq(region *metapb.Region) (*metapb.Peer, err
-	// else we can think the raft server doesn't execute this split command.
-	if leftDetail.Region.RegionEpoch.GetVersion() <= request.Header.RegionEpoch.GetVersion() {
+	reqRegionEncStartKey := encodeRegionStartKey(reqRegion.GetStartKey())
+	reqRegionEncEndKey := encodeRegionSearchKey(reqRegion.GetEndKey())
+	reqRegionPath := makeRegionKey(c.clusterRoot, reqRegion.GetId())
+	reqSearchKey := makeRegionSearchKey(c.clusterRoot, reqRegion.GetEndKey())
+
+	if region == nil {
+		// Find no range after start key, insert directly.
+		ops = append(ops, clientv3.OpPut(reqRegionPath, reqRegionEncEndKey))
+		ops = append(ops, clientv3.OpPut(reqSearchKey, string(regionValue)))
+		return ops, nil
+	}
+
+	// If the request epoch is less than current region epoch, then returns an error.
+	reqRegionEpoch := reqRegion.GetRegionEpoch()
+	regionEpoch := region.GetRegionEpoch()
+	if reqRegionEpoch.GetVersion() < regionEpoch.GetVersion() {
+		// If the request epoch version is less than the current one, return an error.
+		return nil, errors.Errorf("invalid region epoch, request: %v, currenrt: %v", reqRegionEpoch, regionEpoch)
+	} else if reqRegionEpoch.GetVersion() == regionEpoch.GetVersion() {
+		// If the request epoch version is equal to the current one, do nothing.
 		return nil, nil
 	}
 
-	rightDetail, err := c.getRegionDetail(rightRegionID, leader)
-	if err != nil {
-		return nil, errors.Trace(err)
-	}
-
-	return &raft_cmdpb.AdminResponse{
-		CmdType: raft_cmdpb.AdminCmdType_Split.Enum(),
-		Split: &raft_cmdpb.SplitResponse{
-			Left:  leftDetail.Region,
-			Right: rightDetail.Region,
-		},
-	}, nil
-}
-
-func (c *raftCluster) sendRaftCommand(request *raft_cmdpb.RaftCmdRequest, region *metapb.Region) (*raft_cmdpb.RaftCmdResponse, error) {
-	originPeer := request.Header.Peer
+	regionEncStartKey := encodeRegionStartKey(region.GetStartKey())
+	regionEncEndKey := encodeRegionSearchKey(region.GetEndKey())
 
-RETRY:
-	for i := 0; i < maxSendRetry; i++ {
-		resp, err := c.callCommand(request)
-		if err != nil {
-			// We may meet some error, maybe network broken, node down, etc.
-			// We can check later next time.
-			return nil, errors.Trace(err)
+	if reqRegionEncStartKey == regionEncStartKey &&
+		reqRegionEncEndKey == regionEncEndKey {
+		// Seems there is something wrong? Do nothing.
+	} else if regionEncStartKey >= reqRegionEncEndKey {
+		// No range [start, end) in region now, insert directly.
+		ops = append(ops, clientv3.OpPut(reqRegionPath, reqRegionEncEndKey))
+		ops = append(ops, clientv3.OpPut(reqSearchKey, string(regionValue)))
+	} else {
+		// Region overlapped, remove old region, insert new one.
+		// E.g, 1 [a, c) -> 1 [a, b) + 2 [b, c), either new 1 or 2 reports, the region
+		// is overlapped with origin [a, c).
+		regionPath := makeRegionKey(c.clusterRoot, region.GetId())
+		regionSearchKey := makeRegionSearchKey(c.clusterRoot, region.GetEndKey())
+		if regionPath != reqRegionPath {
+			ops = append(ops, clientv3.OpDelete(regionPath))
```

Comment:
seems that we can delete a same key and put it a new value again in the same transaction. 
@xiang90